perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_tcsMails_mu.java

```diff
@@ -0,0 +1,162 @@
+			AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, this.getAttributeDefinition().getName());
+			resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), ((AllAttributesRemovedForUser) message).getUser()));
+
+		}
+
+		return resolvingMessages;
+	}
+
+	/**
+	 * Return true if attribute name is one of affected attributes in this virtual module.
+	 *
+	 * @param nameOfAttribute name of attribute to check
+	 * @return true if attribute is one of affected, false otherwise
+	 */
+	private boolean isAffectedAttribute(String nameOfAttribute) {
+		if(preferredMailFriendlyName.equals(nameOfAttribute)) return true;
+		else if(isMailFriendlyName.equals(nameOfAttribute)) return true;
+		else if(o365MailsFriendlyName.equals(nameOfAttribute)) return true;
+		else if(publicMailsFriendlyName.equals(nameOfAttribute)) return true;
+		else if(privateMailsFriendlyName.equals(nameOfAttribute)) return true;
+
+		return false;
+	}
+
+	/**
+	 * Return email values as sorted set from attribute by name.
+	 *
+	 * It works only for String and List attributes. If attribute has different type of value, it will be logged and skipped.
+	 * If attribute has empty value, it will return empty set.
+	 * If attribute not exists, it will be logged and skipped.
+	 *
+	 * @param sess perun session
+	 * @param user user to get values for
+	 * @param nameOfAttribute name of attribute for which values should be returned
+	 *
+	 * @return sorted set of values of attribute defined by name
+	 */
+	private SortedSet<String> getEmailValues(PerunSessionImpl sess, User user, String nameOfAttribute) {
+		SortedSet<String> valuesToAdd = new TreeSet<>();
+		try {
+			Attribute sourceAttribute = sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, user, nameOfAttribute);
+			if (sourceAttribute.getType().equals(String.class.getName())) {
+				if (sourceAttribute.getValue() != null) valuesToAdd.add(sourceAttribute.valueAsString());
+			} else if (sourceAttribute.getType().equals(ArrayList.class.getName())) {
+				if (sourceAttribute.getValue() != null) valuesToAdd.addAll(sourceAttribute.valueAsList());
+			} else {
+				//unexpected type of value, log it and skip the attribute
+				log.error("Unexpected type of attribute (should be String or ArrayList) {}. It will be skipped.", sourceAttribute);
+			}
+		} catch (AttributeNotExistsException ex) {
+			//we can log this situation and skip the attribute from computing
+			log.warn("When counting value of attribute {} we are missing source attribute {}. Exception: {}. It will be skipped.", this.getAttributeDefinition(), nameOfAttribute, ex);
+		} catch (WrongAttributeAssignmentException ex) {
+			throw new InternalErrorException(ex);
+		}
+
+		return valuesToAdd;
+	}
+
+	@Override
+	public List<String> getStrongDependencies() {
+		List<String> strongDependencies = new ArrayList<>();
+		strongDependencies.add(A_U_D_preferredMail);
```

Comment:
This code is duplicate. Moreover, repeated "add" operations are not necessary.
You can use something like this:
```
		return Arrays.asList(
			A_U_D_preferredMail, 
			A_U_D_ISMail, 
			A_U_D_publicAliasMails, 
			A_U_D_privateAliasMails, 
			A_U_D_o365EmailAddressesMU);
```

Reply:
Changed.